lldb/source/Plugins/Process/elf-core/ProcessElfGpuCore.cpp

```diff
@@ -0,0 +1,169 @@
+//===-- ProcessElfGpuCore.cpp
+//------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <cstdlib>
+
+#include <memory>
+
+#include "ProcessElfGpuCore.h"
+
+#include "lldb/Core/Debugger.h"
+#include "lldb/Core/Module.h"
+#include "lldb/Core/PluginManager.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/Log.h"
+
+using namespace lldb_private;
+using namespace lldb;
+
+llvm::Expected<lldb::TargetSP>
+ProcessElfGpuCore::CreateGpuTarget(lldb_private::Debugger &debugger) {
+
+  // Create target for GPU
+  TargetSP gpu_target_sp;
+  llvm::StringRef exe_path;
+  llvm::StringRef triple;
+  OptionGroupPlatform *platform_options = nullptr;
+
+  Status error(debugger.GetTargetList().CreateTarget(
+      debugger, exe_path, triple, eLoadDependentsNo, platform_options,
+      gpu_target_sp));
+
+  if (error.Fail())
+    return error.ToError();
+
+  if (!gpu_target_sp)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "failed to create GPU target");
+
+  return gpu_target_sp;
+}
+
+llvm::Expected<std::shared_ptr<ProcessElfGpuCore>>
+ProcessElfGpuCore::LoadGpuCore(std::shared_ptr<ProcessElfCore> cpu_core_process,
+                               const FileSpec &core_file) {
+  Log *log = GetLog(LLDBLog::Process);
+  LLDB_LOGF(
+      log,
+      "ProcessElfGpuCore::LoadGpuCore() - Looking for GPU data in core file");
+
+  lldb_private::Debugger &debugger =
+      cpu_core_process->GetTarget().GetDebugger();
+
+  // First, try embedded core plugins (simpler callback mechanism)
+  // These don't require a GPU target upfront
+  std::shared_ptr<ProcessElfGpuCore> gpu_process_sp;
+  for (uint32_t idx = 0; GetEmbeddedCoreCreateCallbackAtIndex(idx) != nullptr;
+       ++idx) {
+    auto create_callback = GetEmbeddedCoreCreateCallbackAtIndex(idx);
+    llvm::StringRef plugin_name = GetEmbeddedCorePluginNameAtIndex(idx);
+
+    LLDB_LOGF(log, "LoadGpuCore() - Trying embedded core plugin: %s",
+              plugin_name.str().c_str());
+
+    gpu_process_sp =
+        create_callback(cpu_core_process, debugger.GetListener(), &core_file);
+
+    if (gpu_process_sp) {
+      LLDB_LOGF(log, "LoadGpuCore() - Embedded plugin %s created process",
+                plugin_name.str().c_str());
+      break;
```

Comment:
we shouldn't break, there can be more than one embedded core file

Reply:
Right, my initial design did not plan to support more than one embedded cores because I do not have a scenario to test about. But I think it makes sense it.